azure-kusto-data/azure/kusto/data/_token_providers.py

```diff
@@ -61,16 +61,23 @@ class TokenProviderBase(abc.ABC):
 
     def __init__(self, kusto_uri: str):
         self._kusto_uri = kusto_uri
-        if kusto_uri is not None:
-            self._scopes = [kusto_uri + ".default" if kusto_uri.endswith("/") else kusto_uri + "/.default"]
-            self._cloud_info = CloudSettings.get_cloud_info()
+
+    def init_cloud(self):
+        if self._kusto_uri is not None:
+            self._cloud_info = CloudSettings.get_cloud_info_for_cluster(self._kusto_uri)
+            resource_uri = self._cloud_info.kusto_service_resource_id
+            if self._cloud_info.login_mfa_required:
+                resource_uri = resource_uri.replace(".kusto.", ".kustomfa.")
+
+            self._scopes = [resource_uri + "/.default"]
 
     def get_token(self):
         """Get a token silently from cache or authenticate if cached token is not found"""
 
         if not self._initialized:
             with TokenProviderBase.lock:
                 if not self._initialized:
+                    self.init_cloud()

```

Comment:
a) You need to be absolutely sure cloud info is not used in the initializers of serviced class
b) I think this is hard constraint, and in most cases it would be served from cache. Consider moving back to __init__

Reply:
Putting it in init is wrong, because:
a. It can potentially take a long time, without the user explictly asking it 2
b. It would block on async flows, which is bad

So that needs to stay, but maybe there is something clever to do in regards to a) and the later context comment

Reply:
I'd wrap this in a private init_once function and call it before Context.
You can have the base call it in Context() and make all derived classes implement context_impl()

Reply:
Is this ok?